src/vmspawn/vmspawn-util.c

```diff
@@ -0,0 +1,341 @@
+/* SPDX-License-Identifier: LGPL-2.1-or-later */
+
+#include "alloc-util.h"
+#include "architecture.h"
+#include "fd-util.h"
+#include "fileio.h"
+#include "json.h"
+#include "log.h"
+#include "macro-fundamental.h"
+#include "memory-util.h"
+#include "path-lookup.h"
+#include "path-util.h"
+#include "sort-util.h"
+#include "string-util-fundamental.h"
+#include "string-util.h"
+#include "recurse-dir.h"
+#include "strv.h"
+#include "vmspawn-util.h"
+#include <stdio.h>
+
+
+OvmfConfig* ovmf_config_free(OvmfConfig *config) {
+        if (!config)
+                return NULL;
+
+        free(config->path);
+        free(config->vars);
+        return mfree(config);
+}
+
+int qemu_check_kvm_support(void) {
+        int r;
+        r = access("/dev/kvm", F_OK);
+        if (r == 0)
+                return true;
+
+        if (errno == ENOENT) {
+                log_debug_errno(errno, "/dev/kvm not found. Not using KVM acceleration.");
+                return false;
+        }
+
+        if (errno == EPERM) {
+                log_debug_errno(errno, "Permission denied to access /dev/kvm. Not using KVM acceleration.");
+                return false;
+        }
+
+        return -errno;
+}
+
+/* holds the information needed to sort the config files into the right order */
+typedef struct OvmfConfigFile {
+        char *name;
+        int priority;
+} OvmfConfigFile;
+
+static inline void ovmf_config_file_free_many(OvmfConfigFile *confs, size_t n_confs) {
+        assert(confs || n_confs == 0);
+
+        FOREACH_ARRAY(conf, confs, n_confs)
+                free(conf->name);
+
+        free(confs);
+}
+
+static int ovmf_config_file_compare(const OvmfConfigFile *a, const OvmfConfigFile *b) {
+        int name_cmp = strcmp(a->name, b->name);
+        return (name_cmp != 0) ? name_cmp : CMP(a->priority, b->priority);
+}
+
+/* holds the data retrieved from the QEMU firmware interop JSON data */
+typedef struct FirmwareData {
+        char **features;
+        char *firmware;
+        char *vars;
+} FirmwareData;
+
+static FirmwareData* firmware_data_new(void) {
+        return new0(FirmwareData, 1);
+}
+
+static FirmwareData* firmware_data_free(FirmwareData *fwd) {
```

Comment:
please make all "deep" destructors handle NULL as param gracefully. i.e. add `if (!fwd) return NULL;` here

Reply:
Fixed in 69a7728f8a.